packages/typegpu-color/src/oklab.ts

```diff
@@ -0,0 +1,346 @@
+        const mdt2 = 6 * m_dt * m_dt * m_;
+        const sdt2 = 6 * s_dt * s_dt * s_;
+
+        const r = 4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s - 1;
+        const r1 = 4.0767416621 * ldt - 3.3077115913 * mdt + 0.2309699292 * sdt;
+        const r2 =
+          4.0767416621 * ldt2 - 3.3077115913 * mdt2 + 0.2309699292 * sdt2;
+
+        const u_r = r1 / (r1 * r1 - 0.5 * r * r2);
+        let t_r = -r * u_r;
+
+        const g = -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s - 1;
+        const g1 =
+          -1.2684380046 * ldt + 2.6097574011 * mdt - 0.3413193965 * sdt;
+        const g2 =
+          -1.2684380046 * ldt2 + 2.6097574011 * mdt2 - 0.3413193965 * sdt2;
+
+        const u_g = g1 / (g1 * g1 - 0.5 * g * g2);
+        let t_g = -g * u_g;
+
+        const b = -0.0041960863 * l - 0.7034186147 * m + 1.707614701 * s - 1;
+        const b1 = -0.0041960863 * ldt - 0.7034186147 * mdt + 1.707614701 * sdt;
+        const b2 =
+          -0.0041960863 * ldt2 - 0.7034186147 * mdt2 + 1.707614701 * sdt2;
+
+        const u_b = b1 / (b1 * b1 - 0.5 * b * b2);
+        let t_b = -b * u_b;
+
+        // @ts-expect-error select doesn't work on number?
+        t_r = select(FLT_MAX, t_r, u_r >= 0);
+        // @ts-expect-error
+        t_g = select(FLT_MAX, t_g, u_g >= 0);
+        // @ts-expect-error
+        t_b = select(FLT_MAX, t_b, u_b >= 0);
+
+        t += min(t_r, min(t_g, t_b));
+      }
+    }
+  }
+
+  return t;
+});
+
+export const gamutClipPreserveChroma = tgpu['~unstable'].fn(
+  [vec3f],
+  vec3f,
+)((lab) => {
+  const L = lab.x;
+  const eps = 0.00001;
+  const C = max(eps, length(lab.yz));
+  const a_ = lab.y / C;
+  const b_ = lab.z / C;
+  const L0 = clamp(L, 0, 1);
+  const cusp = findCusp(a_, b_);
+  const t = clamp(findGamutIntersection(a_, b_, L, C, L0, cusp), 0, 1);
+  const L_clipped = mix(L0, L, t);
+  const C_clipped = t * C;
+
+  return vec3f(L_clipped, C_clipped * a_, C_clipped * b_);
+});
+
+export const gamutClipAdaptiveL05 = tgpu['~unstable']
+  .fn(
+    [vec3f, f32],
+    vec3f,
+  )(
+    // TGSL does not have sign and sqrt yet
+    /* wgsl */ `(lab: vec3f, alpha: f32) -> vec3f {
+      let L = lab.x;
+      let eps = 0.00001;
+      let C = max(eps, length(lab.yz));
+      let a_ = lab.y / C;
+      let b_ = lab.z / C;
+
+      let Ld = L - 0.5;
+      let e1 = 0.5 + abs(Ld) + alpha * C;
+      let L0 = 0.5 * (1 + sign(Ld)*(e1 - sqrt(max(0, e1*e1 - 2 * abs(Ld)))));
+
+      let cusp = findCusp(a_, b_);
+      let t = clamp(findGamutIntersection(a_, b_, L, C, L0, cusp), 0, 1);
+      let L_clipped = mix(L0, L, t);
+      let C_clipped = t * C;
+
+      return vec3f(L_clipped, C_clipped * a_, C_clipped * b_);
+    }`,
+  )
+  .$uses({
+    oklabToLinearRgb,
+    findGamutIntersection,
+    linearRgbToOklab,
+    findCusp,
+  });
+
+export const gamutClipAdaptiveL0cusp = tgpu['~unstable']
+  .fn(
+    [vec3f, f32],
+    vec3f,
+  )(
+    // TGSL does not have sign and sqrt yet
```

Comment:
missing sign and sqrt to convert all functions to TGSL.

Reply:
`std.sqrt` just got merged, and `std.sign` is waiting for review #1205

Reply:
Merged! 🚀